app/main.py

```diff
@@ -1,6 +1,43 @@
-from typing import Callable
+from typing import Callable, Any
 
 
 def cache(func: Callable) -> Callable:
-    # Write your code here
-    pass
+    """
+    decorating only functions that take immutable arguments.
+    cache should print "Getting from cache" when returns stored value
+    and "Calculating new result" when run function with new arguments.
+
+    structure for cache:
+        cache_data =
+        {   "function's name1":
+                {   (tuple1 of arguments): result1,
+                    (tuple2 of arguments): result2,
+                    ...
+                    (tupleN of arguments): resultN,
+                }
+            "function's name2": {...}
+            "function's nameN": {...}
+        }
+    """
+    cache_data = {}
+
+    def wrapper(*func_args, **kwargs) -> Any:
+        function_name = str(func)
```

Comment:
```suggestion
        function_name =  func.__name__
        key = (function_name, func_args, frozenset(kwargs))
```

try to simplify the logic
use `key` for it

don't use loops

Reply:
Ok, done